File: 프로그래머스/0/120861. 캐릭터의 좌표/캐릭터의 좌표.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> solution(vector<string> keyinput, vector<int> board) {
    vector<int> answer = {0, 0};
    
    int maxWidth = board[0] / 2;
    int maxHeight = board[1] / 2;
    
    for (string key : keyinput) {
        if (key == "left" && (abs(answer[0] - 1) <= maxWidth)) {
            answer[0]--;
        } else if (key == "right" && (abs(answer[0] + 1) <= maxWidth)) {
            answer[0]++;
        } else if (key == "up" && (abs(answer[1] + 1) <= maxHeight)) {
            answer[1]++;
        } else if (key == "down" && (abs(answer[1] - 1) <= maxHeight)) {
            answer[1]--;
        }
        
    }
    
    return answer;
}

File: 프로그래머스/0/120861. 캐릭터의 좌표/캐릭터의 좌표.java
import java.util.*;

class Solution {
    int row, col;
    
    public boolean isRangeWithin(int x, int y) {
        return Math.abs(x) <= col && Math.abs(y) <= row;
    }
    
    public int[] solution(String[] keyinput, int[] board) {
        col = board[0] / 2;
        row = board[1] / 2;
        int xPos = 0;
        int yPos = 0;
        
        Map<String, Integer> keys = new HashMap<>();
        keys.put("up", 0);
        keys.put("down", 1);
        keys.put("left", 2);
        keys.put("right", 3);
        
        int[] dy = {1, -1, 0, 0};
        int[] dx = {0, 0, -1, 1};
        
        for (String k : keyinput) {
            int idx = keys.get(k);
            
            int nX = xPos + dx[idx];
            int nY = yPos + dy[idx];
            
            if (isRangeWithin(nX, nY)) {
                xPos = nX;
                yPos = nY;
            }
        }
        
        return new int[]{xPos, yPos};
    }
}